xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,216 @@
+     */
+    public XWikiAttachmentList()
+    {
+        map = new TreeMap<String, XWikiAttachment>();
+    }
+
+    /**
+     * Adds attachment to the list in order of filename.
+     * 
+     * @param attachment XWikiAttachment to add to the list
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean add(XWikiAttachment attachment)
+    {
+        map.put(attachment.getFilename(), attachment);
+        super.clear();
+        super.addAll(map.values());
+        onUpdate();
+        added(attachment);
+        return true;
+    }
+    
+    /**
+     * Adds attachment to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param attachment XWikiAttachment to add to the list
+     * @since 10.0RC1
+     */
+    @Override
+    public void add(int index,XWikiAttachment attachment)
+    {
+        map.put(attachment.getFilename(), attachment);
+        super.clear();
+        super.addAll(map.values());
+        onUpdate();
+        added(attachment);
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public void clear()
+    {
+        super.clear();
+        map.clear();
+        onUpdate();
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean addAll(Collection<? extends XWikiAttachment> c)
+    {
+        onUpdate();
+        for (XWikiAttachment x : c) {
+            map.put(x.getFilename(), x);
+            added(x);
+        }
+        super.clear();
+        super.addAll(map.values());
+        return true;
+
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public XWikiAttachment remove(int index)
+    {
+        XWikiAttachment removedAttachment = map.remove(super.get(index).getFilename());
+        onUpdate();
+        return removedAttachment == null ? null : super.remove(index);
+
+    }
+
+    /**
+     * Removes XWikiAttachment.
+     * 
+     * @param attachment XWikiAttachment to remove.
+     * @return XWikiAttachment that was removed or null if not found
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean remove(Object attachment)
+    {
+        String filename = ((XWikiAttachment)(attachment)).getFilename();
+        XWikiAttachment removedAttachment = map.remove(filename);
+        super.clear();
+        super.addAll(map.values());
+        return removedAttachment==null ? false : true;
+    }
+
+    /**
+     * Adds or replaces attachment with the same filename as the parameter.
+     * 
+     * @param attachment the attachment to add to the list
+     * @return the attachment that was added to the list in order of filename
+     * @since 10.0RC1
+     */
+    public XWikiAttachment set(XWikiAttachment attachment)
+    {
+        map.put(attachment.getFilename(), attachment);
+        super.clear();
+        super.addAll(map.values());
+        return attachment;
```

Comment:
This method should have the same behavior than XWikiDocument#setAttachment instead of make setAttachment workaround it. Also there is no point in returning what you get as parameter.